src/MetaMorpho.sol

```diff
@@ -318,8 +318,27 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
 
         if (length > ConstantsLib.MAX_QUEUE_LENGTH) revert ErrorsLib.MaxQueueLengthExceeded();
 
+        Id[] memory sorted = new Id[](length);
         for (uint256 i; i < length; ++i) {
             if (config[newSupplyQueue[i]].cap == 0) revert ErrorsLib.UnauthorizedMarket(newSupplyQueue[i]);
+            sorted[i] = newSupplyQueue[i];
+        }
+
+        // Gnome sort. O(n) if already sorted.
+        uint256 j;
+        while (j < length) {
+            if (j == 0 || Id.unwrap(sorted[j]) >= Id.unwrap(sorted[j - 1])) {
```

Comment:
I feel like `sorted[j]` & `sorted[j-1]` could be stored as variables not only to be cheaper (max hundred gas) but also to be clearer

Reply:
Like `curr = sorted[j]`, `prev = sorted[j-1]`

Reply:
I think you mean this but let me know if I misunderstood

```solidity
        uint256 j;
        Id cur;
        Id prev;
        while (j < length) {
            if (j==0) {
                j++;
            } else {
                cur = sorted[j];
                prev = sorted[j-1];
                if (Id.unwrap(cur) >= Id.unwrap(prev)) {
                    j++;
                } else {
                    (sorted[j], sorted[j - 1]) = (prev, cur);
                    j--;
                }
            }
        }

        prev = Id.wrap(0);
        for (uint256 i; i < length; ++i) {
            if (Id.unwrap(prev) == Id.unwrap(sorted[i])) revert ErrorsLib.DuplicateMarket(sorted[i]);
            prev = sorted[i];
        }
``` 

it's -23k gas (out of ~350k)

Reply:
Indeed! Not sure it's better in the end

The last for loop is maybe too much too, simply naming variables in the while loop could be enough

Reply:
to clarify, you're saying it may be worth moving duplicate detection inside the while loop. I think it's correct because:

- Some pair of equal elements in the array must end up with both elements adjacent to each other. Take the leftmost such pair.
- If the pair was already adjacent before sorting and never moved, since the sort looks at all initially adjacent pairs once, they will be detected
- If the last move that made the pair adjacent moved the right element of the pair into position, the next comparison is between the elements of the pair (`j--`), so they will be detected
- If the last move that made the pair adjacent moved the left element of the pair into position, the next comparison will go left (`j--`) but the loop will not end until `j == length` and so the elements will compared and detected.

That being said it may not be worth it in gas